Source/Layout/ASLayout.mm

```diff
@@ -216,52 +200,70 @@ - (void)setRetainSublayoutLayoutElements:(BOOL)retainSublayoutLayoutElements
 
 #pragma mark - Layout Flattening
 
+- (BOOL)isFlattened
+{
+  // A layout is flattened if its position is null, and all of its sublayouts are of type displaynode with no sublayouts.
+  if (! ASPointIsNull(_position)) {
+    return NO;
+  }
+  
+  for (ASLayout *sublayout in _sublayouts) {
+    if (ASLayoutIsDisplayNodeType(sublayout) == NO || sublayout->_sublayouts.count > 0) {
+      return NO;
+    }
+  }
+  
+  return YES;
+}
+
 - (ASLayout *)filteredNodeLayoutTree NS_RETURNS_RETAINED
 {
-  if (ASLayoutIsFlattened(self)) {
+  if ([self isFlattened]) {
     // All flattened layouts must have this flag enabled
     // to ensure sublayout elements are retained until the layouts are applied.
     self.retainSublayoutLayoutElements = YES;
     return self;
   }
   
   struct Context {
-    ASLayout *layout;
+    unowned ASLayout *layout;
     CGPoint absolutePosition;
   };
   
   // Queue used to keep track of sublayouts while traversing this layout in a DFS fashion.
   std::deque<Context> queue;
-  for (ASLayout *sublayout in self.sublayouts) {
+  for (ASLayout *sublayout in _sublayouts) {
     queue.push_back({sublayout, sublayout.position});
   }
   
-  NSMutableArray *flattenedSublayouts = [[NSMutableArray alloc] init];
+  auto flattenedSublayouts = [[NSMutableArray<ASLayout *> alloc] init];
   
   while (!queue.empty()) {
     const Context context = std::move(queue.front());
     queue.pop_front();
     
-    ASLayout *layout = context.layout;
-    const NSArray<ASLayout *> *sublayouts = layout.sublayouts;
-    const NSUInteger sublayoutsCount = sublayouts.count;
+    unowned ASLayout *layout = context.layout;
+    const NSUInteger sublayoutsCount = layout->_sublayouts.count;
```

Comment:
Direct ivar access isn't so much about avoiding the message-send, it's about avoiding the retain/release pair. `_sublayouts` is already a `+1` so let's use it.